Source/RapyutaSimulationPlugins/Private/Core/RRNetworkPlayerController.cpp

```diff
@@ -0,0 +1,327 @@
+            {
+                UE_LOG(LogRapyutaCore,
+                       Error,
+                       TEXT("Player [%s] found an entity of matching name %s BUT it's not a pawn. "
+                            "Please recheck [robotname] param value."),
+                       *PlayerName);
+            }
+            break;
+        }
+    }
+#endif
+    return matchingPawn;
+}
+
+void ARRNetworkPlayerController::ServerPossessPawn_Implementation(APawn* InPawn)
+{
+    Possess(InPawn);
+}
+
+void ARRNetworkPlayerController::AcknowledgePossession(APawn* InPawn)
+{
+    // NOTE: [AcknowledgePossession] runs on Client only
+    Super::AcknowledgePossession(InPawn);
+
+    if (false == PlayerName.IsEmpty())
+    {
+        UE_LOG(LogRapyutaCore, Warning, TEXT("Player[%s] AcknowledgePossession %s"), *PlayerName, *InPawn->GetName());
+        // Refer to ARRBaseRobot::CreateROS2Interface() for reasons why it is inited here but not earlier
+        ClientInitPawn(InPawn);
+    }
+}
+
+void ARRNetworkPlayerController::ClientInitPawn_Implementation(AActor* InActor)
+{
+    ARRBaseRobot* robot = Cast<ARRBaseRobot>(InActor);
+    if (robot)
+    {
+        // Instantiate ROS2 interface, of which the instance is not replicated, for each possessed robot
+        verify(nullptr == robot->ROS2Interface);
+        robot->CreateROS2Interface();
+        robot->ROS2Interface->Initialize(robot);
+    }
+
+    if (ARRRobotBaseVehicle* robotVehicle = Cast<ARRRobotBaseVehicle>(InActor))
+    {
+        robotVehicle->InitMoveComponent();
+    }
+}
+
+void ARRNetworkPlayerController::BeginPlay()
+{
+    Super::BeginPlay();
+
+#if !WITH_EDITOR
+    // Set PlayerName to [robotname] if specified as an Sim executor arg
+    FString robotName;
+    if (URRCoreUtils::GetCommandLineArgumentValue<FString>(CMDLINE_ARG_NET_CLIENT_ROBOT_NAME, robotName))
+    {
+        // Remove '"' & '=' in Robot Name
+        robotName = robotName.Replace(TEXT("="), TEXT("")).Replace(TEXT("\""), TEXT(""));
+        if (false == robotName.IsEmpty())
+        {
+            PlayerName = robotName;
+            SimStateClientROS2Node->Namespace = robotName;
+            SetName(robotName);
+            ServerSetPlayerName(robotName);
+        }
+    }
+#endif
+
+    if (PlayerName == URRCoreUtils::PIXEL_STREAMER_PLAYER_NAME)
+    {
+        SetControlRotation(FRotator(-50.f, -90.f, 0));
+    }
+
+    if (IsLocalController())
+    {
+        FTimerManager& timerManager = GetWorld()->GetTimerManager();
+        timerManager.SetTimer(PossessTimerHandle, this, &ARRNetworkPlayerController::WaitToPossessPawn, 1.f, true);
+        timerManager.SetTimer(ClockRequestTimerHandle, this, &ARRNetworkPlayerController::RequestServerTimeUpdate, 5.f, true);
+    }
+}
+
+void ARRNetworkPlayerController::ServerSetPlayerName_Implementation(const FString& InPlayerName)
+{
+    PlayerName = InPlayerName;
+    if (SimStateClientROS2Node)
+    {
+        SimStateClientROS2Node->Namespace = InPlayerName;
```

Comment:
what is this for?

Reply:
originally this was to set the server's name and namespace and let it replicate down, mainly to deal with timing issues from the editor